typhon/display.py

```diff
@@ -42,100 +58,141 @@ class TyphonDisplay(TyphonBase):
 
     parent: QWidget, optional
     """
-    def __init__(self, name=None, image=None, parent=None):
+    # Template types and defaults
+    Q_ENUMS(TemplateTypes)
+    TemplateEnum = TemplateTypes.to_enum()  # For convenience
+    default_templates = dict((_typ.name, os.path.join(ui_dir,
+                                                      _typ.name + '.ui'))
+                             for _typ in TemplateEnum)
+
+    def __init__(self,  parent=None, **kwargs):
+        # Intialize background variable
+        self._use_template = ''
+        self._use_default = False
+        self._last_macros = dict()
+        self._main_widget = None
+        self._template_type = TemplateTypes.detailed_screen
+        # Set this to None first so we don't render
         super().__init__(parent=parent)
-        # Instantiate UI
-        self.ui = uic.loadUi(os.path.join(ui_dir, 'device.ui'), self)
-        # Set Label Names
-        if name:
-            self.title = name
-        # Create child panels
-        self.method_panel = FunctionPanel()
-        self.read_panel = SignalPanel()
-        self.config_panel = SignalPanel()
-        self.misc_panel = SignalPanel()
-        # Add to layout
-        self.read_widget.setLayout(self.read_panel)
-        self.config_widget.setLayout(self.config_panel)
-        self.misc_widget.setLayout(self.misc_panel)
-        self.main_layout.insertWidget(3, self.method_panel,
-                                      0, Qt.AlignHCenter)
-        self.method_panel.hide()
-        # Create PyDMDrawingImage
-        self.image_widget = None
-        if image:
-            self.add_image(image)
-
-    @property
-    def title(self):
-        """Title at the top of panel"""
-        return self.name_label.text()
-
-    @title.setter
-    def title(self, text):
-        self.name_label.setText(text)
+        # Initialize blank UI
+        self.setLayout(QHBoxLayout())
+        self.layout().setContentsMargins(0, 0, 0, 0)
+        # Load template
+        self.load_template()
 
     @property
-    def methods(self):
-        """
-        Methods contained within :attr:`.method_panel`
-        """
-        return self.method_panel.methods
-
-    def add_image(self, path):
+    def current_template(self):
+        """Current template being rendered"""
+        # If a user forces a template we use it
+        if self._use_template:
+            return self._use_template
+        # Search in the last macros, maybe our device told us what to do
+        template_key = self.TemplateEnum(self._template_type).name
+        if not self._use_default and self._last_macros.get(template_key, None):
+            return self._last_macros[template_key]
+        # Otherwise just use the default
+        return self.default_templates[template_key]
+
+    @Property(TemplateTypes)
```

Comment:
What does capital P property do? Is this another weird Qt thing? What an awful API

Reply:
This is what the `qtpy` renames `pyqtproperty` as

Reply:
That's awful